com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/app/discovery/ZigBeeNodeServiceDiscoverer.java

```diff
@@ -195,24 +201,30 @@ private void startDiscovery(Set<NodeDiscoveryTask> newTasks) {
         // Tasks are managed in a queue. The worker thread will only remove the task from the queue once the task is
         // complete. When no tasks are left in the queue, the worker thread will exit.
         synchronized (discoveryTasks) {
+            logger.debug("{}: Node SVC Discovery: starting new tasks {}", node.getIeeeAddress(), newTasks);
+
             // Remove any tasks that we know are not supported by this device
-            if ((!supportsManagementLqi || node.getNodeDescriptor().getLogicalType() == LogicalType.UNKNOWN)
+            if ((!supportsManagementLqi || node.getNodeDescriptor() != null
+                    && node.getNodeDescriptor().getLogicalType() == LogicalType.UNKNOWN)
                     && newTasks.contains(NodeDiscoveryTask.NEIGHBORS)) {
                 newTasks.remove(NodeDiscoveryTask.NEIGHBORS);
             }
-            if ((!supportsManagementRouting || node.getNodeDescriptor().getLogicalType() == LogicalType.UNKNOWN
-                    || node.getNodeDescriptor().getLogicalType() == LogicalType.END_DEVICE)
+            if ((!supportsManagementRouting || node.getNodeDescriptor() != null
+                    && (node.getNodeDescriptor().getLogicalType() == LogicalType.UNKNOWN
+                            || node.getNodeDescriptor().getLogicalType() == LogicalType.END_DEVICE))
                     && newTasks.contains(NodeDiscoveryTask.ROUTES)) {
                 newTasks.remove(NodeDiscoveryTask.ROUTES);
             }
 
             // Make sure there are still tasks to perform
             if (newTasks.isEmpty()) {
-                logger.debug("{}: Node SVC Discovery: has no tasks to perform", node.getIeeeAddress());
+                logger.debug("{}: Node SVC Discovery: has no new tasks to perform", node.getIeeeAddress());
                 return;
             }
 
-            boolean startWorker = discoveryTasks.isEmpty();
+            // Check the current list of tasks to decide if we need to start the worker
+            // This prevents restarting if we didn't add new tasks, which might overload the system
+            boolean startWorker = discoveryTasks.isEmpty() | (futureTask == null);
```

Comment:
I think booleans should not be `or`ed with the bit-wise operator `|` but rather the `||`, so the compiler can complain once one side changes to a non-boolean by accident. Otherwise the compiler will really use the bit-wise or which is not intended here I guess.

Reply:
Thanks.